### DISCUSSION TITLE
Using Webauthn in an Ethereum Smart Contract

### DISCUSSION THREAD

**Discussion by rersozlu** [MasterKale/SimpleWebAuthn#382] [General]
Hi Matthew and fellows, we have reached out to you via e-mail before. Wanted to move this convo to github in order to get advantage of community. 

We're trying to use webauthn as a custom message signer to be used in a smart contract on Ethereum (assume that it uses r1 instead of k1). 

Flow will be as followed: 

1. user will create pvkey - public key pair via webauthn in their device
2. public key will be saved to the smart contract (or we can lets assume its a database)
3. challenge (or transaction hash in web3 case) will be created and sent for signing to user
4. challenge will be signed
5. backend or smart contract will be verifying signature besides with challenge, checking if it returns public key or not (considering it uses secp256r1 curve)

**problem**: client does not directly signs challenge but also clientData and authorizationData. So giving signature and challenge is not enough to retrieve public key but instead we need to construct clientData from scratch. 

**question**: are there any ways to use webauthn prompter on browser or devices to sign a custom message only without sending additional info? if not, should we consider creating clientData manually (assume that there are no libraries to use in web3) ?

**Comment by MasterKale** [comment]
Hello @rersozlu, I don't 100% understand your use case but does Issue #166 sound like what you're trying to achieve?

In PR #172 I updated the `expectedChallenge` option in both `verifyRegistrationResponse()` and `verifyAuthenticationResponse()` to accept a method that can peel apart arbitrary passed in as `challenge` in the corresponding options passed to WebAuthn's `.create()` and `.get()`, and still verify that a challenge was present but also have additional info signed over/contained in the response. That feature was released way back in v4.4.0 and so as long as you're up-to-date with SimpleWebAuthn you'll be able to use that capability.

**Reply by rersozlu** [reply]
I will try to implement this today, to clarify, use case is just to sign challenge with private key and retrieve public key from the signature with r1 curve. 
Webauthn, as far as I understand, does not sign the challenge itself but a combined version of authData and clientData with it. We just want to sign challenge and get the signature of that with r1.

**Reply by rersozlu** [reply]
@MasterKale Tried it, whereas this implementation can lead us to use our backend as a relayer, I still couldn't manage to sign only challenge instead of signing authData and clientData with it. Seems like it is built into the credetentials.create() method natively.

**Reply by MasterKale** [reply]
> use case is just to sign challenge with private key and retrieve public key from the signature with r1 curve

Ah, okay, then I can definitely say that what you're trying to do isn't feasible with WebAuthn. Those other values getting signed over are too vital for the technology's phishing resistance.

I can't think of any other way of signing values like that with WebAuthn... 🤔
